### PR TITLE
Fix SAM.alignlength(record::Record)::Int

### LINKED ISSUE #37: Incorrect calculation of alignment length in SAM.alignlength()
The alignment length is not properly calculated for SAM records using `SAM.alignlength()` also leading to incorrect results with `SAM.rightposition()`.
The problem does not occur in the equivalent BAM accessor functions.

### PR DESCRIPTION
**Pull request by herbstk**
Fix #37 by re-initializing `len` in `SAM.alignlength`.

### CODE REVIEW

**Comment by kescobo** [comment]
Appveyor seems to be failing on nightly, because of test setup (Pkg has changed since 1.0), but that shouldn't matter for this. Were tests failing on the previous version? If not, can you add a test that would look for the issue that this fixes?

**Comment by herbstk** [comment]
To be honest I haven't checked tests, just noticed the issue by accident. From a quick look at it I think the current test set needs a slight modification and then it will accommodate this issue. I will get to it tomorrow.

**Comment by herbstk** [comment]
Ok, @kescobo, after familiarizing myself a bit more with unit testing, I think we would need to change the SAM record file used from [BioFmtSpecimens](https://github.com/BioJulia/BioFmtSpecimens) to test the SAM Reader functionality in BioAlignments.jl. The currently used SAM file for this purpose ([ce#1.sam](https://github.com/BioJulia/BioFmtSpecimens/blob/master/SAM/ce%231.sam)) contains only matching and deletion operations in the CIGAR string, but would also need e.g. insertion or clipping operations to throw an error  when `SAM.rightposition` is tested on the pre-PR version of BioAlignments.jl. This is how I encountered the bug initially because when such operations appear in the SAM.Record's CIGAR field, they erroneously sum into the `alignlength` (and therefore lead to an erroneous `rightposition`). For me the most simple solution was to also test `SAM.rightposition` on the 6th record of file [ce#5b.sam](https://github.com/BioJulia/BioFmtSpecimens/blob/master/SAM/ce%235b.sam), which contains such CIGAR operations. I have also added the same test for `BAM.Records`, although there the `alignlength` is correctly calculated. Anyway, I am not sure if this is the best solution.

**Comment by herbstk** [comment]
In addition, I encountered an inconsistency between the output to STDOUT of `Base.show` for `SAM.Record` and `BAM-Record`, leading to a failure of the latter for the aforementioned 6th Record in file [ce#5b.sam](https://github.com/BioJulia/BioFmtSpecimens/blob/master/SAM/ce%235b.sam) (compare [Base.show(BAM.Record)](https://github.com/herbstk/BioAlignments.jl/blob/645f76a4b00d34098db2f6b55b63035633a45ff9/src/sam/record.jl#L77) and [Base.show(BAM.Record)](https://github.com/herbstk/BioAlignments.jl/blob/645f76a4b00d34098db2f6b55b63035633a45ff9/src/bam/record.jl#L66])). I think I would be able to fix this as well. Should I make another issue and PR?

**Comment by kescobo** [comment]
@herbstk re: file needs - I'll take your word for it. Best thing to do would be to make a PR to BioFmtSpecimens adding an additional file that has the features you need, rather than change the existing one (only because I don't know if other packages rely on that specific file format.

 Re: show methods. Those should definitely be unified. My ideal situation would be to have a single record type across all file formats that uses traits to call the right methods (for everything, not just show) but that's a much bigger topic. If you can unify these show methods, that would be great - I think in a separate PR (no need to make an issue as well, you can just reference this PR)

**Comment by herbstk** [comment]
Regarding the file, I was probably ambiguous. One could add a new file to BioFmtSpecimens and instead test for that one as you suggested. Another option is to use in addition to the currently used test file another one which is already present in BioFmtSpecimens just for the purpose of testing `rightposition`. I did now the latter as this avoids another PR and more extensive changes to the unit tests.

**Comment by herbstk** [comment]
Ok, it is more complicated with fixing `Base.show` for BAM records than I initially thought. I was planning to use `hasX` calls to check for presence or absence of certain fields as it is done for `SAM.Records` but they sometimes behave ambiguously or return the wrong result. Similar issues were risen before in #26 by @jakobnissen  and aparently there would be some refactoring needed for the `BAM.Record` structure and its functions. Currently, I don't have the time to familiarize myself enough with the inner workings of the BAM specs and BioAlignments.jl to get this task done. If one anyway plans to align the `Record` structures for SAM and BAM, this will be much easier to fix. Do you know, what the plans are in that regard? I saw that there is some movement to insulate SAM and BAM file/record handling from BioAlignments.jl with [XAM.jl](https://github.com/BioJulia/XAM.jl) which is probably the better place to start this task.

**Comment by kescobo** [comment]
>I did now the latter as this avoids another PR and more extensive changes to the unit tests. But you would prefer the first solution?

If you found what you need in the files already present in BioFmtSpecimens, great! I was only saying to add an additional file if you need it. Totally fine to use multiple files from that repo in different tests 👍 

>Ok, it is more complicated with fixing Base.show for BAM records than I initially thought. 

No worries - totally understand. In that case, I think open an issue and just let it be for now.

File: test/runtests.jl [hunk 1]
```diff
@@ -1306,6 +1310,10 @@ end
         @test record.tlen == new_record.tlen
         @test record.data == new_record.data
 
+        # rightposition (also implicitly alignlength)
+        records = collect(open(BAM.Reader, joinpath(bamdir, "ce#5b.bam")))
+        @test BAM.rightposition(records[6]) == rightposition(records[6]) == 83
```
Comment (kescobo) [inline test/runtests.jl]: Did you add this same code twice? It's [also here](https://github.com/BioJulia/BioAlignments.jl/pull/38/files#diff-fce720c43af3c52c862fd7451c7374b8R1152-R1155)

Comment (kescobo) [reply test/runtests.jl]: Oh, I see it's sam and bam, nevermind!

**Comment by CiaranOMara** [comment]
These commits were included in XAM.

- https://github.com/BioJulia/XAM.jl/commit/8d14563f9d77cb0996ce6ab7839dc29d932e238f 
- https://github.com/BioJulia/XAM.jl/commit/6a33ac06a7a325de347950b8bb6a8f5beec782bb

**Pull request closed without merging**